lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,247 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module PowerShell
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @stdin, @stdout, @ps_process = Open3.popen2(cmd)
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code, timeout_ms = 300 * 1000)
+        output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        output_ready_event = self.class.create_event(output_ready_event_name)
+
+        # always need a trailing newline to ensure PowerShell parses code
+        code = <<-CODE
```

Comment:
Is there a way we caan break this out to a separate file or method? This is getting long and could get problematic in future @Iristyle @ferventcoder @glennsarti